src/event_processor/processor/locking/locks.clj

```diff
@@ -1,28 +1,43 @@
 (ns event-processor.processor.locking.locks
-  (:require [hugsql.core :as hugsql]
-            [event-processor.utils.logging :as log]))
-
-(declare get-lock release-lock)
-(hugsql/def-db-fns "event_processor/processor/locking/locks.sql")
+  (:require [event-processor.utils.logging :as log]))
 
 (defmacro with-lock
-  "Acquire an advisory lock on the database, and run function, then release the lock.
-   Block if lock is taken."
+  "Try to acquire a Postgres advisory lock on the database.
+  If successful, run `body`, otherwise skip."
   [database lock-id & body]
-  `(try
-     (log/log-debug
-       {:lock-id ~lock-id}
-       "Acquiring lock")
+  `(let [db# (:handle ~database)
+         datasource# (:datasource db#)
+         connection# (.getConnection datasource#)]
      (try
-       (get-lock (:handle ~database) {:lock_id ~lock-id})
-       (catch Exception exception#
-         (throw (ex-info "Error getting locks"
-                  {:lock_id ~lock-id :lock-error true}
-                  exception#))))
-     ~@body
-     (finally
-       (do
-         (log/log-debug
-           {:lock-id ~lock-id}
-           "Releasing lock")
-         (release-lock (:handle ~database) {:lock_id ~lock-id})))))
+       (log/log-debug
+         {:lock-id ~lock-id}
+         "Acquiring lock")
+       (let [get-lock-statement#
+             (.prepareStatement
+               connection#
+               "SELECT pg_try_advisory_lock(?)")
+
+             _# (.setInt get-lock-statement# 1 ~lock-id)
+             result-set# (.executeQuery get-lock-statement#)
+
+             _# (.first result-set#)
+             acquired# (.getBoolean result-set# 1)
+
+             _# (.close result-set#)]
+
+         (when acquired#
+           ~@body))
+
+       (finally
+         (do
+           (log/log-debug
+             {:lock-id ~lock-id}
+             "Releasing lock")
+           (let [release-lock-statement#
```

Comment:
Should this still run the release-lock even when acquired is false?

Reply:
fixed - only release when the lock is obtained or where there's an explicit exception